### PR TITLE
Hnsw links storage

### PR DESCRIPTION
**Pull request by IvanPleshkov**
New data structure for containing graph links. It uses less RAM memory and it's a middle point of memmap for hnsw graph links.

RAM usage after simple `bfb` test with 1M points:
`dev: 703MB`
`branch: 653MB`

`bfb` build and search results (1M points):
```
DEV:
Index ready in 175.2995453 seconds
p95 search time: 0.000497951
p99 search time: 0.000562454
Max search time: 0.00870695
```

```
Branch:
Index ready in 174.302510991 seconds
p95 search time: 0.000490513
p99 search time: 0.000555676
Max search time: 0.008915356
```

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/index/hnsw_index/graph_layers.rs [hunk 1]
```diff
@@ -241,20 +240,24 @@ impl GraphLayers {
...
+        self.links.num_points()
     }
 
     pub fn point_level(&self, point_id: PointOffsetType) -> usize {
-        self.links_layers[point_id as usize].len() - 1
+        self.links.point_level(point_id)
     }
 
     pub fn merge_from_other(&mut self, other: GraphLayers) {
+        self.max_level = std::cmp::max(self.max_level, other.max_level);
```
Comment (generall) [inline lib/segment/src/index/hnsw_index/graph_layers.rs:251]: De-compressing flatten representation might be expensive in the scenario, when we need to merge a lot of small graphs (the common scenario, actually). Let's only use merge on regular graph representation and compress only the final version of it

Comment (IvanPleshkov) [reply lib/segment/src/index/hnsw_index/graph_layers.rs]: fixed

File: lib/segment/src/index/hnsw_index/graph_layers_builder.rs [hunk 2]
```diff
@@ -655,18 +693,24 @@ mod tests {
         }
         let graph_layers = graph_layers_builder.into_graph_layers();
 
-        let number_layers = graph_layers.links_layers.len();
+        let number_layers = graph_layers.links.num_points();
```
Comment (generall) [inline lib/segment/src/index/hnsw_index/graph_layers_builder.rs:696]: It is number of points or number of layers?

Comment (IvanPleshkov) [reply lib/segment/src/index/hnsw_index/graph_layers_builder.rs]: In dev `graph_layers.links_layers.len()` is definitely a number of points and there is a mistake in the text message. The number of layers is present in the next `eprintln`.

Fixed message text and variable name.

**Review by generall — APPROVED** [review]

**Pull request merged**